### ISSUE TITLE
CORNER NET: AssertionError: RandomCenterCropPad needs the input image of dtype np.float32, please set "to_float32=True" in "LoadImageFromFile" pipeline

### ISSUE THREAD

**Issue by ravising-h** [open-mmlab/mmdetection#3581]
**Checklist**
- [x] I have searched related issues but cannot get the expected help. 
- [x]  The bug has not been fixed in the latest version. 

**Describe the bug**
I have trained cornernet model and 
`AssertionError: RandomCenterCropPad needs the input image of dtype np.float32, please set "to_float32=True" in "LoadImageFromFile" pipeline` and this error shows while inference. (to_float is True in config)

**Reproduction**

```
img = mmcv.imread('/content/drive/My Drive/R-FCN/data/coco/JPEGImages/renamed_4210.jpg')

model.cfg = cfg
result = inference_detector(model, img)
show_result_pyplot(model, img, result)
```
2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?
XML DATASET (CUSTOM  only added class) 

**Environment**

```
2020-08-17 18:09:55,706 - mmdet - INFO - Environment info:
------------------------------------------------------------
sys.platform: linux
Python: 3.6.9 (default, Jul 17 2020, 12:50:27) [GCC 8.4.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 10.1, V10.1.243
GPU 0: Tesla K80
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.5.1+cu101
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.5 Product Build 20190808 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 

TorchVision: 0.6.1+cu101
OpenCV: 4.1.2
MMCV: 1.0.5
MMDetection: 2.3.0+6acf6be
MMDetection Compiler: GCC 7.5
MMDetection CUDA Compiler: 10.1
------------------------------------------------------------

2020-08-17 18:09:55,707 - mmdet - INFO - Distributed training: False
2020-08-17 18:09:56,095 - mmdet - INFO - Config:
checkpoint_config = dict(interval=10)
log_config = dict(interval=50, hooks=[dict(type='TextLoggerHook')])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
model = dict(
    type='CornerNet',
    backbone=dict(
        type='HourglassNet',
        downsample_times=5,
        num_stacks=2,
        stage_channels=[256, 256, 384, 384, 384, 512],
        stage_blocks=[2, 2, 2, 2, 2, 4],
        norm_cfg=dict(type='BN', requires_grad=True)),
    neck=None,
    bbox_head=dict(
        type='CornerHead',
        num_classes=1,
        in_channels=256,
        num_feat_levels=2,
        corner_emb_channels=1,
        loss_heatmap=dict(
            type='GaussianFocalLoss', alpha=2.0, gamma=4.0, loss_weight=1),
        loss_embedding=dict(
            type='AssociativeEmbeddingLoss',
            pull_weight=0.25,
            push_weight=0.25),
        loss_offset=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1)))
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile', to_float32=True),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PhotoMetricDistortion',
        brightness_delta=32,
        contrast_range=(0.5, 1.5),
        saturation_range=(0.5, 1.5),
        hue_delta=18),
    dict(
        type='RandomCenterCropPad',
        crop_size=(511, 511),
        ratios=(0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3),
        test_mode=False,
        test_pad_mode=None,
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Resize', img_scale=(511, 511), keep_ratio=False),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]
test_pipeline = [
    dict(type='LoadImageFromFile', to_float32=True),
    dict(
        type='MultiScaleFlipAug',
        scale_factor=1.0,
        flip=True,
        transforms=[
            dict(type='Resize'),
            dict(
                type='RandomCenterCropPad',
                crop_size=None,
                ratios=None,
                border=None,
                test_mode=True,
                test_pad_mode=['logical_or', 127],
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(
                type='Collect',
                keys=['img'],
                meta_keys=('filename', 'ori_shape', 'img_shape', 'pad_shape',
                           'scale_factor', 'flip', 'img_norm_cfg', 'border'))
        ])
]
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    train=dict(
        type='MyDataset',
        ann_file='/content/drive/My Drive/R-FCN/data/coco/train_dent.txt',
        img_prefix='/content/drive/My Drive/R-FCN/data/coco/',
        pipeline=[
            dict(type='LoadImageFromFile', to_float32=True),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(
                type='PhotoMetricDistortion',
                brightness_delta=32,
                contrast_range=(0.5, 1.5),
                saturation_range=(0.5, 1.5),
                hue_delta=18),
            dict(
                type='RandomCenterCropPad',
                crop_size=(511, 511),
                ratios=(0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3),
                test_mode=False,
                test_pad_mode=None,
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Resize', img_scale=(511, 511), keep_ratio=False),
            dict(type='RandomFlip', flip_ratio=0.5),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
        ]),
    val=dict(
        type='MyDataset',
        ann_file='/content/drive/My Drive/R-FCN/data/coco/test_dent.txt',
        img_prefix='/content/drive/My Drive/R-FCN/data/coco/',
        pipeline=[
            dict(type='LoadImageFromFile', to_float32=True),
            dict(
                type='MultiScaleFlipAug',
                scale_factor=1.0,
                flip=True,
                transforms=[
                    dict(type='Resize'),
                    dict(
                        type='RandomCenterCropPad',
                        crop_size=None,
                        ratios=None,
                        border=None,
                        test_mode=True,
                        test_pad_mode=['logical_or', 127],
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(
                        type='Collect',
                        keys=['img'],
                        meta_keys=('filename', 'ori_shape', 'img_shape',
                                   'pad_shape', 'scale_factor', 'flip',
                                   'img_norm_cfg', 'border'))
                ])
        ]),
    test=dict(
        type='MyDataset',
        ann_file='/content/drive/My Drive/R-FCN/data/coco/test.txt',
        img_prefix='/content/drive/My Drive/R-FCN/data/coco/',
        pipeline=[
            dict(type='LoadImageFromFile', to_float32=True),
            dict(
                type='MultiScaleFlipAug',
                scale_factor=1.0,
                flip=True,
                transforms=[
                    dict(type='Resize'),
                    dict(
                        type='RandomCenterCropPad',
                        crop_size=None,
                        ratios=None,
                        border=None,
                        test_mode=True,
                        test_pad_mode=['logical_or', 127],
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(
                        type='Collect',
                        keys=['img'],
                        meta_keys=('filename', 'ori_shape', 'img_shape',
                                   'pad_shape', 'scale_factor', 'flip',
                                   'img_norm_cfg', 'border'))
                ])
        ]))
train_cfg = None
test_cfg = dict(
    corner_topk=100,
    local_maximum_kernel=3,
    distance_threshold=0.5,
    score_thr=0.05,
    max_per_img=100,
    nms_cfg=dict(type='soft_nms', iou_threshold=0.5, method='gaussian'))
optimizer = dict(type='Adam', lr=0.0005)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.3333333333333333,
    step=[180])
total_epochs = 11
dataset_type = 'MyDataset'
data_root = '/content/drive/My Drive/R-FCN/data/coco/'
evaluation = dict(interval=12, metric='bbox')
work_dir = './work_dirs/cornernet_hourglass104_mstest_10x5_210e_coco'
gpu_ids = range(0, 1)

2020-08-17 18:10:03,484 - mmdet - INFO - Start running, host: root@ead17a162f1a, work_dir: /content/mmdetection/work_dirs/cornernet_hourglass104_mstest_10x5_210e_coco
2020-08-17 18:10:03,484 - mmdet - INFO - workflow: [('train', 1)], max: 11 epochs
2020-08-17 18:15:00,067 - mmdet - INFO - Epoch [1][50/375]	lr: 1.993e-04, eta: 6:42:50, time: 5.931, data_time: 0.079, memory: 10550, det_loss: 1892.8123, off_loss: 19.1445, pull_loss: 696.7372, push_loss: 0.0067, loss: 2608.7008, grad_norm: 373508.4643
2020-08-17 18:19:53,047 - mmdet - INFO - Epoch [1][100/375]	lr: 2.327e-04, eta: 6:35:29, time: 5.860, data_time: 0.009, memory: 10550, det_loss: 54.5773, off_loss: 15.0010, pull_loss: 534.7616, push_loss: 0.0219, loss: 604.3618, grad_norm: 29908.6013
2020-08-17 18:24:46,235 - mmdet - INFO - Epoch [1][150/375]	lr: 2.660e-04, eta: 6:29:52, time: 5.864, data_time: 0.009, memory: 10550, det_loss: 46.9078, off_loss: 12.4506, pull_loss: 1293.7027, push_loss: 0.0590, loss: 1353.1201, grad_norm: 62648.4193
2020-08-17 18:29:39,138 - mmdet - INFO - Epoch [1][200/375]	lr: 2.993e-04, eta: 6:24:31, time: 5.858, data_time: 0.009, memory: 10550, det_loss: 30.5830, off_loss: 8.5783, pull_loss: 938.9489, push_loss: 0.1025, loss: 978.2128, grad_norm: 54262.1950
2020-08-17 18:34:32,058 - mmdet - INFO - Epoch [1][250/375]	lr: 3.327e-04, eta: 6:19:22, time: 5.858, data_time: 0.009, memory: 10550, det_loss: 74.8030, off_loss: 7.5447, pull_loss: 2966.4483, push_loss: 0.2229, loss: 3049.0189, grad_norm: 33954.0571
2020-08-17 18:39:24,955 - mmdet - INFO - Epoch [1][300/375]	lr: 3.660e-04, eta: 6:14:18, time: 5.858, data_time: 0.009, memory: 10550, det_loss: 26.4749, off_loss: 5.6188, pull_loss: 140.9628, push_loss: 0.1617, loss: 173.2182, grad_norm: 5608.6533
2020-08-17 18:44:17,955 - mmdet - INFO - Epoch [1][350/375]	lr: 3.993e-04, eta: 6:09:18, time: 5.860, data_time: 0.009, memory: 10550, det_loss: 16.7364, off_loss: 3.1022, pull_loss: 34.8091, push_loss: 0.2154, loss: 54.8631, grad_norm: 1957.3584

```
Training continued
**Error traceback**
If applicable, paste the error trackback here.
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-12-1d152aae2563> in <module>()
      2 
      3 model.cfg = cfg
----> 4 result = inference_detector(model, img)
      5 show_result_pyplot(model, img, result)

4 frames
/content/mmdetection/mmdet/apis/inference.py in inference_detector(model, img)
     94     # prepare data
     95     data = dict(img=img)
---> 96     data = test_pipeline(data)
     97     data = collate([data], samples_per_gpu=1)
     98     if next(model.parameters()).is_cuda:

/content/mmdetection/mmdet/datasets/pipelines/compose.py in __call__(self, data)
     38 
     39         for t in self.transforms:
---> 40             data = t(data)
     41             if data is None:
     42                 return None

/content/mmdetection/mmdet/datasets/pipelines/test_time_aug.py in __call__(self, results)
    103                 _results['flip'] = flip
    104                 _results['flip_direction'] = direction
--> 105                 data = self.transforms(_results)
    106                 aug_data.append(data)
    107         # list of dict to dict of list

/content/mmdetection/mmdet/datasets/pipelines/compose.py in __call__(self, data)
     38 
     39         for t in self.transforms:
---> 40             data = t(data)
     41             if data is None:
     42                 return None

/content/mmdetection/mmdet/datasets/pipelines/transforms.py in __call__(self, results)
   1511         img = results['img']
   1512         assert img.dtype == np.float32, (
-> 1513             'RandomCenterCropPad needs the input image of dtype np.float32,'
   1514             ' please set "to_float32=True" in "LoadImageFromFile" pipeline')
   1515         h, w, c = img.shape

AssertionError: RandomCenterCropPad needs the input image of dtype np.float32, please set "to_float32=True" in "LoadImageFromFile" pipeline
```

**Comment by tianyuandu** [comment]
Hi, I noticed the bug appeared when you call the function `inference_detector`. And I'm wondering will the same error came out while testing after an epoch training?

**Comment by ldz666666** [comment]
i came up with the save error, althought i have changed the function in /mmdetection/mmdet/datasets/pipelines/loading.py, speifying class LoadImageFromFile to_float32=True

**Comment by tianyuandu** [comment]
> i came up with the save error, althought i have changed the function in /mmdetection/mmdet/datasets/pipelines/loading.py, speifying class LoadImageFromFile to_float32=True

Did you call the same function `inference_detector` or use `demo/image_demo.py`?

**Comment by ravising-h** [comment]
@tianyuandu No testing script works ok when I call --out
error show in --show-dir

**Comment by ravising-h** [comment]
I have found Loading Image Happens in LoadImage Func of Inference script and not from loading in mmdet.datasets .

**Comment by tianyuandu** [comment]
> I have found Loading Image Happens in LoadImage Func of Inference script and not from loading in mmdet.datasets .

Yes, the Loading pipeline in `inference_detector` is hard-coded and maybe misalign to config. I'll fix this later.

**Comment by tianyuandu** [comment]
> I have found Loading Image Happens in LoadImage Func of Inference script and not from loading in mmdet.datasets .

Yes, the Loading pipeline in `inference_detector` is hard-coded and maybe misalign to config. I'll fix this later.

> @tianyuandu No testing script works ok when I call --out
> error show in --show-dir
> I did not evaluated after every epochs.

Did you use `tools/test.py` with `--out` and `--show-dir` and this raise errors? Could you show me your detail command?

**Comment by ldz666666** [comment]
> > i came up with the save error, althought i have changed the function in /mmdetection/mmdet/datasets/pipelines/loading.py, speifying class LoadImageFromFile to_float32=True
> 
> Did you call the same function `inference_detector` or use `demo/image_demo.py`?

Yes , i just called the inference_detector function in inference.py in mmdetection/mmdet/api folder

**Comment by tianyuandu** [comment]
@ravising-h @ldz666666 the bug of `inference_detector` is fixed by #3611, and changes were merged into master. Please pull the code and see whether the problem is solved, thank you!
